---
title: "RNASeq analysis"
output: html_document
author: Hong Zheng
date: Sep. 2017
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=4, fig.height=4, fig.path='Figs/',
                      echo=FALSE, warning=FALSE, message=FALSE)
```

## Download reference and prepare index
```{r prepare, echo = T, eval=FALSE}
#######
# These steps are performed in linux/bash, in reference/ directory
# Script location: reference/run_download_index.sh
# Kallisto and Salmon are from pre-built Docker images. Scripts for building the images can be found in https://github.com/zhengh42/Dockerfiles
#######

#######
# Download from Gencode release M15
#######
wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M15/GRCm38.primary_assembly.genome.fa.gz
wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M15/gencode.vM15.annotation.gtf.gz
wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M15/gencode.vM15.transcripts.fa.gz
gzip -d *.gz

#######
# Stats of Gencode release M15
#######
# Number of genes: 52550
less gencode.vM15.annotation.gtf | awk '$3=="gene"' | wc -l
# Number of transcript: 131100
less gencode.vM15.annotation.gtf | awk '$3=="transcript"' | wc -l
less gencode.vM15.transcripts.fa | grep '^>' | wc -l
# Get gene name and type
less gencode.vM15.annotation.gtf  | awk '$3=="gene"' | cut -f9 | awk '{print $2,$4,$6}' | sed 's/[";]//g;s/\s\+/\t/g' | sed '1i geneid\tgenetype\tgenename' > gene.info

#######
# Generate transcript to gene mapping file
#######
less gencode.vM15.transcripts.fa | egrep '^>' | sed 's/^>//' | awk 'OFS="\t"{print $0,$0}' | sed 's/|/\t/;s/|/\t/' | awk 'OFS="\t"{print $4,$2}' > tx2gene.txt

transcripts_fasta=gencode.vM15.transcripts.fa
#######
# Kallisto index
#######
docker run -v $PWD:/mnt zhengh42/kallisto:0.43.1 kallisto index -i /mnt/M15.gencode.kallisto.idx /mnt/$transcripts_fasta

#######
# Salmon index
#######
docker run -v $PWD:/mnt zhengh42/salmon:0.8.2 salmon index -i /mnt/M15.gencode.salmon.idx -t /mnt/$transcripts_fasta
```

## QC and transcript-level expression quantification

```{r preprocess, echo = T, eval=FALSE}
#######
# These steps are performed in linux/bash, in scripts/ directory
# Trim-galore is from pre-built Docker images. Scripts for building the images can be found in https://github.com/zhengh42/Dockerfiles
# Script location: scripts/run_indiv.sh
# How to run:
# ./run_indiv.sh sham1 Raw_sham1
# ./run_indiv.sh sham2 Raw_sham2
# ./run_indiv.sh sham3 Raw_sham3
# ./run_indiv.sh WT_Ang1 Raw_WT_Ang1
# ./run_indiv.sh WT_Ang2 Raw_WT_Ang2
# ./run_indiv.sh WT_Ang3 Raw_WT_Ang3
#######

#######
# Arguments setup
#######
id=$1
id2=$2
raw_dir=/srv/gevaertlab/data/Hong/RNASeq/hb/Reads
work_dir=/srv/gevaertlab/data/Hong/RNASeq/hb/RNASeq_MM_HB/
ref_dir=$work_dir/reference

#######
# QC with trim-galore and fastqc
#######
docker run -v $raw_dir:/home zhengh42/trim-galore:0.4.4  \
        trim_galore -q 15  --stringency 3 --gzip --length 15 --paired 
        /home/$id/${id2}_1.fq.gz /home/$id/${id2}_2.fq.gz --fastqc --output_dir /home 
        1> ../logs/$id.trim_galore.log 2>&1

#######
# Get the stand-specific information of the reads
# In the log file of salmon output: Automatically detected most likely library type as ISR
#######
zcat $raw_dir/${id2}_1_val_1.fq.gz | head -n 400000 | gzip > $raw_dir/${id2}_test_1.fq.gz
zcat $raw_dir/${id2}_2_val_2.fq.gz | head -n 400000 | gzip > $raw_dir/${id2}_test_2.fq.gz
docker run -v $raw_dir:/home/seq -v $ref_dir:/home/ref -v $work_dir/results/salmon:/home/out  \
        zhengh42/salmon:0.8.2 \
        salmon quant -i /home/ref/M15.gencode.salmon.idx -l A \
        -o /home/out/$id.test -1 /home/seq/${id2}_test_1.fq.gz -2 /home/seq/${id2}_test_2.fq.gz \
        1> ../logs/$id.salmon.log 2>&1

#######
# Get transcript level expression estimate using Kallisto
# Since the library type is ISR, --rf-stranded argument is specified
# More information on the library type and strand information, see http://fishycat.netlify.com/en/2017/08/strandness_in_rnaseq/
#######
docker run -v $raw_dir:/home/seq -v $ref_dir:/home/ref -v $work_dir/results/kallisto:/home/out \
        zhengh42/kallisto:0.43.1 \
        kallisto quant -i /home/ref/M15.gencode.kallisto.idx \
        -o /home/out/$id /home/seq/${id2}_1_val_1.fq.gz /home/seq/${id2}_2_val_2.fq.gz  \
        -b 100 --rf-stranded --fusion  1> ../logs/$id.kallisto.log 2>&1
```

## Differential expression analysis

```{r rsetup, echo = T}
#######
# The following steps are performed in R, in scripts/ directory
#######

#######
# Load required packages, set up directory, read in files
#######
library(readr)
library(tximport)
require("knitr")
require(DESeq2) # version 1.16.1
require(IHW)
require(dplyr)
require(ComplexHeatmap)
require(vsn)
require(RColorBrewer)
require(ggplot2)

#work_dir<- "/srv/gevaertlab/data/Hong/RNASeq/hb/RNASeq_MM_HB/"
#opts_knit$set(root.dir = work_dir)
#sampleID <- scan(paste0(work_dir,"sampleID"),what="character",quiet=TRUE)
#geneinfo<-read.table(paste0(work_dir,"reference/gene.info"),head=T,sep = "\t")
```

```{r load_save_worksapce}
load(file=".RData")
#save.image(file=".RData")
```

```{r kallisto, echo = T, eval=FALSE}
#######
# Get gene-level expression from transcript level results of Kallisto
#######
files.kallisto <- file.path(paste0(work_dir,"results/kallisto"),sampleID,"abundance.tsv")
all(file.exists(files.kallisto))
names(files.kallisto)<- sampleID
tx2gene<-read.table(paste0(work_dir,"reference/tx2gene.txt"))
Kallisto.txim <- tximport(files.kallisto,type="kallisto",tx2gene = tx2gene)
write.csv(Kallisto.txim,file=paste0(work_dir, "results/kallisto/Kallisto.txim.txt"),row.names = T)
```

### 1. DESeq2

```{r DESeq2,  echo = T, eval=F}
#######
# Read in data and pre-filtering
#######
sampleTable=data.frame(condition = factor(rep(c("wt-sham","wt-ang"),each=3)))
rownames(sampleTable) = colnames(Kallisto.txim$counts)
dds<- DESeqDataSetFromTximport(Kallisto.txim,sampleTable,~condition)
dds$condition <- factor(dds$condition, levels = c("wt-sham","wt-ang"))
dim(dds) #[1] 52550     6
# Only genes with at least two counts across the samples were kept
dds <- dds[ rowSums(counts(dds)) > 1, ]
dim(dds) #[1] 31043     6

#######
# Generate DESeq object
#######
dds <- DESeq(dds)
```


```{r data_transformation, echo=T, eval=T}
#######
# Different approaches of tranformation. DESeq uses regularized log transformation as default.
#######
# Shifted logarithm transformation, log2(n + 1), has elevated standard deviation in the lower count range.
ntd <- normTransform(dds)
meanSdPlot(assay(ntd))

# Regularized log transformation, less elevated standard deviation in the lower count range.
rld <- rlog(dds, blind=FALSE)
meanSdPlot(assay(rld))

# Variance stabilizing transformation.It may over-correct standard deviation and mask true differences due to the experimental conditions.
vsd <- varianceStabilizingTransformation(dds, blind=FALSE)
meanSdPlot(assay(vsd))
```

```{r heatmap,echo=T, eval=T}
#######
# Data quality assessment by sample clustering and visualization
#######
# Sample clustering based on first 400 highly-expressed genes
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:400]

p<- Heatmap(assay(rld)[select,],
        clustering_distance_rows = "euclidean", clustering_distance_columns = "euclidean",
        clustering_method_rows  = "average",clustering_method_columns  = "average",
        show_row_names =F,
        name="distance",
        column_names_gp = gpar(fontsize = 9)
             )
draw(p,newpage = T)

### Heatmap of the sample-to-sample distances
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
colnames(sampleDistMatrix) <- NULL

colors <- colorRampPalette( rev(brewer.pal(9, "Oranges")) )(255)
p<- Heatmap(sampleDistMatrix,
        clustering_distance_rows = "euclidean", clustering_distance_columns = "euclidean",
        clustering_method_rows  = "average",clustering_method_columns  = "average",
        col=colors,
        name="Distance",
        column_names_gp = gpar(fontsize = 9)
             )
draw(p,newpage = T)

# PCA plot
pcaData <- plotPCA(rld, intgroup=c("condition"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=condition,shape=name)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

__QC summary__

From the clustering and visualization plots above, we can see that sham1 and sham2 always cluster together, as well as Ang1 and Ang2. However, instead of clustering with sham group, sham3 always clusters with Ang group. Is there any problem with this sample?

To accurately measure fold change, two versions of differential expression analysis was performed, one with sham3 and one with out.

__Differential expression analysis with all samples__

```{r DE_analysis,  echo = T, eval=T}
#######
# Differential expression analysis
#######
res <- results(dds)
# Get moderated and shrunken log2 fold changes. No effect on p values
resLFC <- lfcShrink(dds, coef=2, res=res)
```

```{r MAPlot,  echo = T, eval=T}
#######
# MA plot
#######
plotMA(res, ylim=c(-2,2))
plotMA(resLFC, ylim=c(-2,2))
```

```{r DESeq2_DE_genes, echo = T,  eval=T}
#######
# Choose top differentially-expressed genes
#######
# By default, DESeq performs independent filtering based on the mean of normalized counts for each gene, optimizing the number of genes which will have an adjusted p value below a given FDR cutoff (default 0.1). Using alpha to change the threshold: results(dds, alpha=0.2)
summary(res)

# Filtering by Independent Hypothesis Weighting
resIHW <- results(dds, filterFun=ihw)
summary(resIHW)
```

There are `r sum(res$padj<0.1,na.rm=T)` genes with adjusted p value less than 0.1 using default independent filtering.

There are `r sum(resIHW$padj<0.1,na.rm=T)` genes with adjusted p value less than 0.1 using independent hypothesis weighting filtering.

```{r DESeq2_output,  echo = T, eval=T}
######
# Write results
#######
DESeq.gene.out <- as.data.frame(res)
DESeq.gene.out$log2FoldChange_lfc <- resLFC$log2FoldChange
DESeq.gene.out$padj_IHW <- resIHW$padj
DESeq.gene.out$geneid <- rownames(res)
DESeq.geneanno.out <- merge(x=DESeq.gene.out,y=geneinfo,by = "geneid", all.x = TRUE)

Kallisto.txim.abundance <- as.data.frame(Kallisto.txim$abundance)
colnames(Kallisto.txim.abundance) <- gsub("^","TPM.",colnames(Kallisto.txim.abundance))
Kallisto.txim.abundance$geneid<-rownames(Kallisto.txim.abundance)
Kallisto.txim.counts <- as.data.frame(Kallisto.txim$counts)
colnames(Kallisto.txim.counts) <- gsub("^","counts.",colnames(Kallisto.txim.counts))
Kallisto.txim.counts$geneid<-rownames(Kallisto.txim.counts)

DESeq.geneanno.out<-merge(x=DESeq.geneanno.out,y=Kallisto.txim.abundance,by = "geneid", all.x = TRUE)
DESeq.geneanno.out<-merge(x=DESeq.geneanno.out,y=Kallisto.txim.counts,by = "geneid", all.x = TRUE)
DESeq.geneanno.out.ordered<-DESeq.geneanno.out[order(DESeq.geneanno.out$padj_IHW),]

# What are the types of the genes?
table(DESeq.geneanno.out.ordered$genetype)

# Get lncRNA genes
lncrna<-c("lincRNA","antisense_RNA","sense_intronic","sense_overlapping","TEC")
DESeq.geneanno.out.ordered.lncRNA <- DESeq.geneanno.out.ordered %>% filter(genetype %in% lncrna)

write.table(DESeq.geneanno.out.ordered.lncRNA,file = "DESeq.geneanno.out.ordered.lncRNA.txt",sep="\t",quote = F,row.names = F)
write.table(DESeq.geneanno.out.ordered,file = "DESeq.geneanno.out.ordered.txt",sep="\t",quote = F,row.names = F)
```


__Differential expression analysis without sham3__

```{r Kallisto_a,  echo = T, eval=F}
# Start with Kallisto data import
files.kallisto_a <- files.kallisto[-which(names(files.kallisto) %in% c("sham3") )]
Kallisto.txim_a <- tximport(files.kallisto_a,type="kallisto",tx2gene = tx2gene)
```

```{r DESeq_a,  echo = T, eval=F}
# DESeq
sampleTable_a <- sampleTable[-which(rownames(sampleTable)  %in% c("sham3")),,F]
dds_a<- DESeqDataSetFromTximport(Kallisto.txim_a,sampleTable_a,~condition)

dds_a$condition <- factor(dds_a$condition, levels = c("wt-sham","wt-ang"))
dim(dds_a) #[1] 52550     6
# Only genes with at least two counts across the samples were kept
dds_a <- dds_a[ rowSums(counts(dds_a)) > 1, ]
dim(dds_a) #[1] 30796     5

dds_a <- DESeq(dds_a)
res_a <- results(dds_a)
resLFC_a <- lfcShrink(dds_a, coef=2, res=res_a)
```

```{r MAPlot_a,  echo = T, eval=T}
#######
# MA plot
#######
plotMA(res_a, ylim=c(-2,2))
plotMA(resLFC_a, ylim=c(-2,2))
```

```{r DESeq2_degenes_a, echo = T,  eval=T}
# Choose top differentially-expressed genes
summary(res_a)

# Filtering by Independent Hypothesis Weighting
resIHW_a <- results(dds_a, filterFun=ihw)
summary(resIHW_a)
```

There are `r sum(res_a$padj<0.1,na.rm=T)` genes with adjusted p value less than 0.1 using default independent filtering.

There are `r sum(resIHW_a$padj<0.1,na.rm=T)` genes with adjusted p value less than 0.1 using independent hypothesis weighting filtering.

```{r DESeq2_output_a,  echo = T, eval=T}
# Write results
DESeq.gene.out_a <- as.data.frame(res_a)
DESeq.gene.out_a$log2FoldChange_lfc <- resLFC_a$log2FoldChange
DESeq.gene.out_a$padj_IHW <- resIHW_a$padj
DESeq.gene.out_a$geneid <- rownames(res_a)
DESeq.geneanno.out_a <- merge(x=DESeq.gene.out_a,y=geneinfo,by = "geneid", all.x = TRUE)

DESeq.geneanno.out_a<-merge(x=DESeq.geneanno.out_a,y=Kallisto.txim.abundance,by = "geneid", all.x = TRUE)
DESeq.geneanno.out_a<-merge(x=DESeq.geneanno.out_a,y=Kallisto.txim.counts,by = "geneid", all.x = TRUE)
DESeq.geneanno.out.ordered_a<-DESeq.geneanno.out_a[order(DESeq.geneanno.out_a$padj_IHW),]

# What are the types of the genes?
table(DESeq.geneanno.out.ordered_a$genetype)

# Get lncRNA genes
lncrna<-c("lincRNA","antisense_RNA","sense_intronic","sense_overlapping","TEC")
DESeq.geneanno.out.ordered.lncRNA_a <- DESeq.geneanno.out.ordered_a %>% filter(genetype %in% lncrna)

write.table(DESeq.geneanno.out.ordered.lncRNA_a,file = "DESeq.geneanno.out.ordered.lncRNA.wosham3.txt",sep="\t",quote = F,row.names = F)
write.table(DESeq.geneanno.out.ordered,file = "DESeq.geneanno.out.ordered.wosham3.txt",sep="\t",quote = F,row.names = F)
```


__Easy report__
```{r easy_report, echo = T, eval=F}
#######
# This report is based on DESeq analysis of all samples, default filtering. See ./reports for details.
#######
des2Report <- HTMLReport(shortName = 'RNAseq_analysis_with_DESeq2',
  title = 'RNA-seq analysis of differential expression using DESeq2',
  reportDirectory = "./reports")
publish(dds,des2Report, pvalueCutoff=0.1,
  annotation.db="org.Mm.eg.db", factor = colData(dds)$condition,
  reportDir="./reports")
finish(des2Report)
```

__Note__

What's in the output file? 

- geneid  
- log2FoldChange, default fold change output by DESeq2.
- pvalue  
- padj, adjusted p value    
- log2FoldChange_lfc, moderated and shrunken log2 fold changes, more conservative (and accurate?)
- padj_IHW, adjusted p value of independent hypothesis weighting (smaller, which results in more differentially-expressed genes)      
- genetype        
- genename 
- TPM.sample, the transcript per million of the gene in the sample
- counts.sample, the read counts of the gene in the sample

